staging/src/k8s.io/client-go/scale/util.go

```diff
@@ -0,0 +1,156 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package scale
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/discovery"
+	scalescheme "k8s.io/client-go/scale/scheme"
+	scaleautoscaling "k8s.io/client-go/scale/scheme/autoscalingv1"
+	scaleextint "k8s.io/client-go/scale/scheme/extensionsint"
+	scaleext "k8s.io/client-go/scale/scheme/extensionsv1beta1"
+)
+
+// ScaleKindResolver knows about the relationship between
+// resources and the GroupVersionKind of their scale subresources.
+type ScaleKindResolver interface {
+	// ScaleForResource returns the GroupVersionKind of the
+	// scale subresource for the given GroupVersionResource.
+	ScaleForResource(resource schema.GroupVersionResource) (scaleVersion schema.GroupVersionKind, err error)
+}
+
+// discoveryScaleResolver is a ScaleKindResolver that uses
+// a DiscoveryInterface to associate resources with their
+// scale-kinds
+type discoveryScaleResolver struct {
+	discoveryClient discovery.ServerResourcesInterface
+
+	subresMap map[schema.GroupVersionResource]schema.GroupVersion
+	mu        sync.RWMutex
+}
+
+func (r *discoveryScaleResolver) generateKindMap() error {
+	resourceLists, err := r.discoveryClient.ServerResources()
+	if err != nil {
+		return fmt.Errorf("unable to update scale kinds: %v")
+	}
+
+	subresMap := make(map[schema.GroupVersionResource]schema.GroupVersion)
+
+	// first, find the scale subresources that we care about
+	for _, resourceList := range resourceLists {
+		groupVer, err := schema.ParseGroupVersion(resourceList.GroupVersion)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("unable to update scale kinds for API group %s: %v", groupVer.String(), err))
+			continue
+		}
+
+		for _, resource := range resourceList.APIResources {
+			resourceParts := strings.SplitN(resource.Name, "/", 2)
+			if resource.Kind != "Scale" || len(resourceParts) != 2 || resourceParts[1] != "scale" {
+				// skip non-scale resources
+				continue
+			}
+
+			mainResource := groupVer.WithResource(resourceParts[0])
+			scaleGV := groupVer
+			if resource.Group != "" && resource.Version != "" {
+				scaleGV = schema.GroupVersion{
+					Group:   resource.Group,
+					Version: resource.Version,
+				}
+			}
+
+			subresMap[mainResource] = scaleGV
+		}
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.subresMap = subresMap
+
+	return nil
+}
+
+func (r *discoveryScaleResolver) ScaleForResource(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
+	r.mu.RLock()
+	gv, exists := r.subresMap[resource]
+	if !exists {
+		// retry on misses in case we have an update
```

Comment:
Can you build a non-caching version, then build a caching version on top of it?  This is a little hard for me to read and the complexity can be built up in another pull.

Reply:
sure, can-do

Reply:
hmm... this is a little weirder than I initially thought, since a non-cached version would probably use different code paths entirely (i.e. it wouldn't use the "get-all-at-once" API).

Reply:
> hmm... this is a little weirder than I initially thought, since a non-cached version would probably use different code paths entirely (i.e. it wouldn't use the "get-all-at-once" API).

Yeah, that's the bit I don't like.